### PR TITLE
Avoid a deadlock between SwiftLanguageRuntimeImpl::SetupReflection() 

### PR DESCRIPTION
**Pull request by adrian-prantl**
@adrian-prantl
Avoid a deadlock between SwiftLanguageRuntimeImpl::SetupReflection()  …
fa05ca0
and SwiftLanguageRuntimeImpl::ModulesDidLoad()

SetupABIBit() iterates of the Target's images and thus needs to
acquire that ModuleList's lock. We need to acquire this before
locking m_add_module_mutex, since ModulesDidLoad can also be
called from a place where that lock is already held:
```
+   lldb_private::DynamicLoaderDarwin::AddModulesUsingImageInfos()
+     lldb_private::ModuleList::AppendIfNeeded()
+       lldb_private::Target::NotifyModuleAdded()
+         lldb_private::Target::ModulesDidLoad()
```
rdar://72858062

### CODE REVIEW

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
Note that we now may call SetupABIBit multiple times, but that's safe (just not particularly efficient).

**Comment by jimingham** [comment]
LGTM

File: lldb/source/Target/SwiftLanguageRuntime.cpp [hunk 1]
```diff
@@ -107,19 +107,18 @@ static ModuleSP findRuntime(Process &process, RuntimeKind runtime_kind) {
...
+    if (image && ((runtime_kind == RuntimeKind::Swift &&
+                   IsModuleSwiftRuntime(process, *image)) ||
+                  (runtime_kind == RuntimeKind::ObjC && image &&
+                   objc_runtime->IsModuleObjCLibrary(image)))) {
+      runtime_image = image;
+      return false;
+    }
+    return true;
+  });
+  return runtime_image;
```
Comment (kastiglione) [inline lldb/source/Target/SwiftLanguageRuntime.cpp]: are these changes NFC restructuring?

**Review by kastiglione — APPROVED** [review]

File: lldb/source/Target/SwiftLanguageRuntime.cpp (hunk 1 above)
Comment (jimingham) [reply lldb/source/Target/SwiftLanguageRuntime.cpp]: Avoiding the ModuleList copy should be NFC.  The logic part also looks the same.  

The current code doesn't need to check `image` again in the RuntimeKind::ObjC clause.  

I'm not sure I'm in favor of this rewriting.  It does unify the return, but OTOT this is an order of magnitude harder to read than the old version IMO...

Comment (adrian-prantl) [reply lldb/source/Target/SwiftLanguageRuntime.cpp]: Yes, see the commit message for the first of the two commits.

Comment (adrian-prantl) [reply lldb/source/Target/SwiftLanguageRuntime.cpp]:
```
Avoid copying the entire module list (NFC).  …
fa3181a
This does the exact same amount of locking as the previous version,
it's just slightly more efficient to use ForEach().
```

Comment (adrian-prantl) [reply lldb/source/Target/SwiftLanguageRuntime.cpp]: These refactorings shall serve as proof that I actually read through the code ;-)

Comment (kastiglione) [reply lldb/source/Target/SwiftLanguageRuntime.cpp]: I see now, thanks!

Comment (kastiglione) [reply lldb/source/Target/SwiftLanguageRuntime.cpp]: Like Jim, I find the conditions in original form easier to read/understand.

Comment (jimingham) [reply lldb/source/Target/SwiftLanguageRuntime.cpp]: I'm not objecting to the ForEach use, to be clear.  Just combining the conditionals like that.

**Comment by kastiglione** [comment]
@swift-ci test macOS Platform

**Comment by adrian-prantl** [comment]
Addressed review feedback — thanks!

**Comment by adrian-prantl** [comment]
@swift-ci test

**Pull request merged**